compiler/rustc_ast/src/mut_visit.rs

```diff
@@ -78,79 +86,84 @@ pub trait MutVisitor: Sized {
     // forget to add handling for it.
 
     fn visit_crate(&mut self, c: &mut Crate) {
-        noop_visit_crate(c, self)
+        walk_crate(self, c)
     }
 
     fn visit_meta_list_item(&mut self, list_item: &mut NestedMetaItem) {
-        noop_visit_meta_list_item(list_item, self);
+        walk_meta_list_item(self, list_item);
     }
 
     fn visit_meta_item(&mut self, meta_item: &mut MetaItem) {
-        noop_visit_meta_item(meta_item, self);
+        walk_meta_item(self, meta_item);
     }
 
     fn visit_use_tree(&mut self, use_tree: &mut UseTree) {
-        noop_visit_use_tree(use_tree, self);
+        walk_use_tree(self, use_tree);
     }
 
-    fn flat_map_foreign_item(&mut self, ni: P<ForeignItem>) -> SmallVec<[P<ForeignItem>; 1]> {
-        noop_flat_map_item(ni, self)
+    fn visit_foreign_item(&mut self, ni: P<ForeignItem>) -> SmallVec<[P<ForeignItem>; 1]> {
+        walk_item(self, ni, None)
```

Comment:
Instead of renaming flat_map to visit, should we have both? A visit_* version that takes a &mut and the flat_map version that calls it and returns a singleton smallvec?

Reply:
I'll pull in the commits from https://github.com/rust-lang/rust/pull/127371, they were perf neutral after all

Reply:
Acutally... hmm... I don't like the runtime asserts for producing multiple items. I'll have to think about how we can make it so implementors can override either `visit_foo` or `flat_map_foo` without there being problems. The `walk_bar` functions must invoke `flat_map_foo`, so that they can pick up on more/fewer items being produced. So `walk_flat_map_foo` needs to invoke `visit_foo`, not `walk_foo`. Otherwise overriding `visit_foo` will do nothing.

Reply:
I think I should address this in a separate PR, as it needs some fiddling to get perf right

Reply:
I removed the renames and kept the `flat_map` names